vendor/bat-native-ads/src/bat/ads/internal/security/conversions/conversions_util.cc

```diff
@@ -0,0 +1,122 @@
+
+std::vector<uint8_t> Base64ToBytes(const std::string& value_base64) {
+  std::string value_as_string;
+  base::Base64Decode(value_base64, &value_as_string);
+
+  const std::string hex_encoded =
+      base::HexEncode(value_as_string.data(), value_as_string.size());
+
+  std::vector<uint8_t> bytes;
+  base::HexStringToBytes(hex_encoded, &bytes);
+
+  return bytes;
+}
+
+}  // namespace
+
+base::Optional<VerifiableConversionEnvelopeInfo> EncryptAndEncode(
+    const VerifiableConversionInfo& verifiable_conversion) {
+  const std::string message = verifiable_conversion.id;
+  const std::string public_key_base64 = verifiable_conversion.public_key;
+
+  if (message.length() < kVacMessageMinLength ||
+      message.length() > kVacMessageMaxLength) {
+    return base::nullopt;
+  }
+
+  if (!IsLatinAlphaNumeric(message)) {
+    return base::nullopt;
+  }
+
+  // Protocol requires at least 2 trailing zero-padding bytes
+  std::vector<uint8_t> plaintext(message.begin(), message.end());
+  plaintext.insert(plaintext.end(), kVacCipherTextLength - plaintext.size(), 0);
+  DCHECK_EQ(kVacCipherTextLength, plaintext.size());
+
+  const std::vector<uint8_t> public_key = Base64ToBytes(public_key_base64);
+  if (public_key.size() != kCryptoBoxPublicKeyBytes) {
+    return base::nullopt;
+  }
+
+  const KeyPairInfo ephemeral_key_pair = GenerateBoxKeyPair();
+
+  const std::vector<uint8_t> nonce = GenerateNonce();
+
+  const std::vector<uint8_t> padded_ciphertext =
+      Encrypt(plaintext, nonce, public_key, ephemeral_key_pair.secret_key);
+
+  // The receiving TweetNaCl.js client does not require padding
+  const std::vector<uint8_t> ciphertext(
+      padded_ciphertext.begin() + kCryptoBoxZeroBytes, padded_ciphertext.end());
+
+  VerifiableConversionEnvelopeInfo envelope;
+  envelope.algorithm = kAlgorithm;
+  envelope.ciphertext = base::Base64Encode(ciphertext);
+  envelope.ephemeral_public_key =
+      base::Base64Encode(ephemeral_key_pair.public_key);
+  envelope.nonce = base::Base64Encode(nonce);
+
+  if (!envelope.IsValid()) {
+    return base::nullopt;
+  }
+
+  return envelope;
+}
+
+std::string DecodeAndDecrypt(const VerifiableConversionEnvelopeInfo envelope,
+                             const std::string& advertiser_secret_key_base64) {
+  std::string message;
+  if (!envelope.IsValid()) {
+    return message;
+  }
+
+  std::vector<uint8_t> advertiser_secret_key =
+      Base64ToBytes(advertiser_secret_key_base64);
+  std::vector<uint8_t> nonce = Base64ToBytes(envelope.nonce);
+  std::vector<uint8_t> ciphertext = Base64ToBytes(envelope.ciphertext);
+  std::vector<uint8_t> ephemeral_public_key =
+      Base64ToBytes(envelope.ephemeral_public_key);
+
+  // API requires 16 leading zero-padding bytes
+  ciphertext.insert(ciphertext.begin(), kCryptoBoxZeroBytes, 0);
+
+  std::vector<uint8_t> plaintext =
+      Decrypt(ciphertext, nonce, ephemeral_public_key, advertiser_secret_key);
+  message = (const char*)&plaintext.front();
```

Comment:
Is there no error checking for decryption? Or is the error passed via the Decrypt func itself? I would like to see this fail hard if possible.

Reply:
I've added some checking, but note this is only used in the unittest and never called by any business logic